langchain4j-anthropic/src/main/java/dev/langchain4j/model/anthropic/AnthropicClient.java

```diff
@@ -71,17 +93,161 @@ AnthropicCreateMessageResponse createMessage(AnthropicCreateMessageRequest reque
+                        handler.onNext(text);
+
+                    }
+                }
+            }
+
+            private void handleContentBlockDelta(AnthropicStreamingData data) {
+                if (data.delta != null && "text_delta".equals(data.delta.type)) {
+                    String text = data.delta.text;
+                    if (isNotNullOrEmpty(text)) {
+                        currentContentBuilder.append(text);
+                        handler.onNext(text);
+
+                    }
+                }
+            }
+
+            private void handleContentBlockStop() {
+                contents.add(currentContentBuilder.toString());
+                currentContentBuilder = new StringBuffer();
+            }
+
+            private void handleMessageDelta(AnthropicStreamingData data) {
+                if (data.delta != null) {
+                    AnthropicDelta delta = data.delta;
+                    if (delta.stopReason != null) {
+                        this.stopReason = delta.stopReason;
+                    }
+                }
+                if (data.usage != null) {
+                    handleUsage(data.usage);
+                }
+            }
+
+            private void handleMessageStop() {
+                Response<AiMessage> response = Response.from(
+                        AiMessage.from(String.join("\n", contents)),
+                        new TokenUsage(inputTokenCount.get(), outputTokenCount.get()),
+                        toFinishReason(stopReason)
+                );
+                handler.onComplete(response);
+            }
+
+            private void handleError(String dataString) {
+                handler.onError(new AnthropicHttpException(null, dataString));
+            }
+
+            @Override
+            public void onFailure(EventSource eventSource, Throwable t, okhttp3.Response response) {
+                if (logResponses) {
+                    log.debug("onFailure()", t);
+                }
+
+                if (t != null) {
+                    handler.onError(t);
+                } else if (response != null) {
+                    try {
+                        if (response.body() != null) {
+                            handler.onError(new AnthropicHttpException(response.code(), response.body().string()));
+                        }
+                    } catch (IOException e) {
+                        handler.onError(new AnthropicHttpException(response.code(), null));
+                    }
+                }
+            }
+
+            @Override
+            public void onClosed(EventSource eventSource) {
+                if (logResponses) {
+                    log.debug("onClosed()");
+                }
+            }
+        };
+
+        Call<ResponseBody> call = anthropicApi.streamCreateMessage(apiKey, version, request);
+        EventSources.createFactory(okHttpClient).newEventSource(call.request(), eventSourceListener);
```

Comment:
I'm sorry, I'm not familiar with `EventSources`, and I don't quite understand the need for a `volatile` variable, and the `synchronized` list, so I may not be able to contribute effectively here.

Reply:
No worries!

This code will be rewritten soon anyway as we will decouple integration modules from okhttp and make it possible to use any other http client of choice.
volatile/synchronized is needed to ensure that the code stays thread safe IF the underlying http client handles events on different threads